doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,1170 @@
+    that.
+  - Attempts were also made to use a window-broker-content architecture, with
+    the broker process having a static CLSID in the registry, and having the
+    window and content processes at mixed elevation levels `CoCreateInstance`
+    that broker. This however _also_ did not work across elevation levels. This
+    may be due to a lack of Packaged COM support for mixed elevation levels.
+    Even if this approach did end up working, we would still need to be
+    responsible for securing the elevated windows so that an unelevated attacker
+    couldn't hijack a content process and trigger unexpected code in the window
+    process. We didn't feel confident that we could properly secure this channel
+    either.
+
+### Monarch and Peasant Processes
+
+With the current design, it's easy to connect many content processes to a window
+process, and move those content processes easily between windows. However, we
+want to make sure that it's also possible to communicate between these
+processes. What if we want to have only a single WT instance, so that whenever
+WT is run, it creates a tab in the existing window? What if you want to run a
+commandline in a given WT window?
+
+In addition to the concept of Window and Content Processes, we'll also be
+introducing another type of categorization for window processes. These are
+"Monarch" and "Peasant" processes. This will allow for the coordination across
+the various windows by the Monarch.
+
+There will only ever be one monarch process at a given time, and every other WT
+window process is a peasant process. However, we want this system to be
+redundant, so that if the monarch ever dies, one of the remaining peasant
+processes can take over for it. The new monarch will be chosen at random, so
+we'll call this a probabilistic elective monarchy.
+
+Essentially, the probabilistic elective monarchy will work in the following way:
+
+1. We'll introduce a WinRT class (for the purpose of this doc we'll call it
+  `Monarch`), with a unique GUID.
+2. When any window process starts up, it'll first try to register as the server
+   for the `Monarch` WinRT class.
+   - The OS will allow subsequent processes to successfully register as the
+    server for `Monarch` objects, but when someone tries to `create_instance` a
+    `Monarch`, the OS will always create the `Monarch` from the first process to
+    register.
+3. After registering as a server for `Monarch`s, attempt to create a `Monarch`
+   using `winrt::create_instance`.
+4. Using that `Monarch`, ask it for it's PID.
+   - If that PID is the same as the PID of the current process, then the window
+     process knows that it is the monarch.
+   - If that PID is some other process, then we know that we're not currently
+     the monarch.
+     - If we don't currently have an ID assigned to us, then ask the `Monarch`
+       to assign one to us.
+     - If we do have an ID (from a previous monarch), then let the new monarch

```

Comment:
what if the monarch rejects the ID?

Reply:
Probaby just assign a new one? Interesting. This hasn't happened in practice.